pkg/loop/client.go

```diff
@@ -0,0 +1,154 @@
+package loop
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/smartcontractkit/chainlink-relay/pkg/logger"
+)
+
+var _ grpc.ClientConnInterface = (*atomicClient)(nil)
+
+type atomicClient struct {
+	cc atomic.Pointer[grpc.ClientConn]
+}
+
+func (a *atomicClient) store(cc *grpc.ClientConn) { a.cc.Store(cc) }
+
+func (a *atomicClient) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	return a.cc.Load().Invoke(ctx, method, args, reply, opts...)
+}
+
+func (a *atomicClient) NewStream(ctx context.Context, desc *grpc.StreamDesc, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
+	return a.cc.Load().NewStream(ctx, desc, method, opts...)
+}
+
+var _ grpc.ClientConnInterface = (*clientConn)(nil)
+
+// newClientFn returns a new client connection id to dial, and a set of resource dependencies to close.
+type newClientFn func(context.Context) (id uint32, deps resources, err error)
+
+// clientConn is a [grpc.ClientConnInterface] backed by a [*grpc.ClientConn] which can be recreated and swapped out
+// via the provided [newClientFn].
+type clientConn struct {
+	*lggrBroker
+	newClient newClientFn
+	name      string
+
+	mu   sync.RWMutex
+	deps resources
+	cc   *grpc.ClientConn
+}
+
+func (lb *lggrBroker) newClientConn(name string, newClient newClientFn) *clientConn {
+	return &clientConn{
+		lggrBroker: lb.named(name),
+		newClient:  newClient,
+		name:       name,
+	}
+}
+
+func (c *clientConn) Invoke(ctx context.Context, method string, args interface{}, reply interface{}, opts ...grpc.CallOption) error {
+	c.mu.RLock()
+	cc := c.cc
+	c.mu.RUnlock()
+
+	if cc == nil {
+		cc = c.refresh(ctx, nil)
+	}
+	for cc != nil {
+		err := cc.Invoke(ctx, method, args, reply, opts...)
+		if err == nil || !isErrTerminal(err) {
+			return err
+		}
+		cc = c.refresh(ctx, cc)
```

Comment:
Because each iteration represents a terminal error that requires reconnecting. It can only go as fast as we can relaunch and initialize everything, so it is indirectly throttled. We currently get stuck, by design, to support methods w/o error signatures and to not impose upstream refactoring - but we may want to relax this in the future. (FWIW context.Context is still respected here)

Reply:
I mean we keep retrying until we can return a 'real' response, or the system is shutting down.
If we give up eventually that assumes the conditions can't/won't change. Even with a static docker image, we don't necessarily know that is true, and we are trying to preserve the non-LOOPP behavior of boot-looping when the node crashes.